apollo-router-core/src/cache.rs

```diff
@@ -0,0 +1,174 @@
+use crate::{CacheResolver, CacheResolverError};
+use derivative::Derivative;
+use futures::lock::Mutex;
+use lru::LruCache;
+use std::cmp::Eq;
+use std::collections::HashMap;
+use std::fmt;
+use std::hash::Hash;
+use tokio::sync::broadcast::{self, Sender};
+
+/// A result from a cache get().
+pub type CacheResult<V> = Result<V, CacheResolverError>;
+
+/// A caching map optimised for slow value resolution.
+///
+/// The CachingMap hold values in an LruCache. Values are loaded into the cache on a cache miss and
+/// the cache relies on the resolver to provide values. There is no way to manually remove, update
+/// or otherwise invalidate a cache value at this time. Values will be evicted from the cache once
+/// the cache_limit is reached.
+#[derive(Derivative)]
```

Comment:
No we don't need but it's a nice to have and it costs 5 lines. I'm fine with it and implementing Debug as much as possible makes the API nicer to use.

So I'm not against removing it. Slight preference for keeping it.

Reply:
It's required because:
```
apollo-router-core/src/query_planner/caching_query_planner.rs

#[derive(Debug)]
pub struct CachingQueryPlanner<T: QueryPlanner> {
    cm: CachingMap<QueryKey, Arc<QueryPlan>>,
...
```

Reply:
That's not required. We can use derivative there if we want

Reply:
why would having the cache implement Debug make the API nice to use?

Reply:
I don't have strong feelings about this but I think it's better to control the output of Debug here rather than in every consumer of this struct.

Reply:
> why would having the cache implement Debug make the API nice to use?

Because people don't need to use derivative when they want to wrap the object. They can just `derive(Debug)` and it works immediately.

Reply:
> I think it's better to control the output of Debug here rather than in every consumer of this struct.

Less lines using derivative (also less opportunities to f up the implementation). Though I don't mind if you do want to implement Debug yourself.

Reply:
I do not think there are cases where we would need to print the cache, especially if it's only to show the cache limit. it should not have to implement Debug at all

Reply:
Ok let's remove the debug impl then because I don't have strong opinion on this.

> So I'm not against removing it. Slight preference for keeping it.

Reply:
I think we've missed something here though. We have to have Debug on CachingQueryPlanner right now, because it is collected in a caching span. (#[instrument]). So, it might be the right thing to not have a Debug on the cache, but to avoid it we have to exclude it in all Debug consumers, so right now, it's the right answer (I think). But, if we remove Debug requirements upstream, we can remove this one at the same time.

Maybe putting the above comment on the source would be helpful?

Reply:
Oh!! That's because it is on the trait `QueryPlanner`. I don't mind removing this trait boundary. Then on `ApolloRouter` you can use derivative to exclude the field from the debug output.

Reply:
> We have to have Debug on CachingQueryPlanner right now, because it is collected in a caching span. (#[instrument]).

that sounds like something we do not want to have (cf the other perf issues we had with #[instrument])

Reply:
I think the use of "skip_all" as recommended in the ADR on telemetry will sort out the telemetry perf problem when we implement it.

I'm trying to minimise the "blast radius" of this PR and so I'd like to avoid making changes outside of the required by the PR changes. So, I want to avoid modifying QueryPlanner/ApolloRouter/<whatever else>... and contain the changes to the minimum set. Keeping this change for now and fixing later and documenting the ability to remove later in a comment seems the cleanest approach.

Reply:
This discussion has been extracted and deferred into issue: #244